File: SAMUeL/python-backend/manuscript_generator.py
from __future__ import annotations

import subprocess
from pathlib import Path
from typing import Any


def write_file(path: Path, text: str) -> None:
    path.write_text(text, encoding="utf-8")


def generate_manuscript(manuscript_dir: Path, stats: dict[str, Any]) -> None:
    manuscript_dir.mkdir(parents=True, exist_ok=True)
    write_file(
        manuscript_dir / "introduction.tex",
        "\\section{Introduction}\nSAMUeL enables annotation-guided SAM segmentation for microscopy.\n",
    )
    write_file(
        manuscript_dir / "methods.tex",
        "\\section{Architecture}\n"
        "The system includes:\n"
        "\\begin{itemize}\n"
        "\\item QuPath Java extension\n"
        "\\item FastAPI Python backend\n"
        "\\item Tile-based inference\n"
        "\\item Mask reconstruction\n"
        "\\end{itemize}\n",
    )
    write_file(
        manuscript_dir / "results.tex",
        "\\section{Usage Statistics}\n"
        f"Tiles processed: {stats.get('tiles', 0)}\\\\\n"
        f"Objects generated: {stats.get('objects', 0)}\\\\\n",
    )
    write_file(
        manuscript_dir / "discussion.tex",
        "\\section{Technical Details}\n"
        "SAM models supported: vit_h, vit_l, vit_b\\\\\n"
        "Processing: Overlapping tile strategy\\\\\n"
        "Output: QuPath-compatible objects\n",
    )
    write_file(
        manuscript_dir / "supplementary.tex",
        "\\section*{Configuration}\n"
        "Backend: FastAPI on port 8000\\\\\n"
        "Dependencies: PyTorch, segment-anything, OpenCV\\\\\n"
        "Integration: HTTP REST API\n",
    )
    write_file(
        manuscript_dir / "references.bib",
        "@article{sam2023,\n  title={Segment Anything},\n  author={Kirillov, A. and others},\n  year={2023}\n}\n"
        "@software{qupath,\n  title={QuPath},\n  author={Bankhead, P. and others},\n  year={2017}\n}\n",
    )
    write_file(
        manuscript_dir / "manuscript.tex",
        "\\documentclass{article}\n"
        "\\usepackage[margin=1in]{geometry}\n"
        "\\begin{document}\n"
        "\\title{SAMUeL Technical Documentation}\n\\maketitle\n"
        "\\input{introduction.tex}\n"
        "\\input{methods.tex}\n"
        "\\input{results.tex}\n"
        "\\input{discussion.tex}\n"
        "\\input{supplementary.tex}\n"
        "\\bibliographystyle{plain}\n\\bibliography{references}\n"
        "\\end{document}\n",
    )
    subprocess.run(
        ["pdflatex", "manuscript.tex"],
        cwd=str(manuscript_dir),
        check=False,
        capture_output=True,
        text=True,
    )

File: SAMUeL/src/main/java/org/samuel/manuscript/ManuscriptGenerator.java
package org.samuel.manuscript;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Map;

/**
 * Writes a manuscript scaffold and compiles it via pdflatex.
 */
public class ManuscriptGenerator {

    public void generate(Path manuscriptDir, Map<String, Object> stats) throws IOException, InterruptedException {
        Files.createDirectories(manuscriptDir);
        write(manuscriptDir.resolve("introduction.tex"), """
                \\section{Introduction}
                SAMUeL integrates Segment Anything in QuPath for annotation-prompted microscopy segmentation.
                """);
        write(manuscriptDir.resolve("methods.tex"), """
                \\section{Methods}
                The workflow used annotation prompts, tile-based extraction, FastAPI model inference, and mask fusion.
                """);
        write(manuscriptDir.resolve("results.tex"), """
                \\section{Results}
                Processed tiles: %d\\\\
                Generated objects: %d\\\\
                """.formatted(stats.getOrDefault("tiles", 0), stats.getOrDefault("objects", 0)));
        write(manuscriptDir.resolve("discussion.tex"), """
                \\section{Discussion}
                Tile overlap reduces border artifacts while preserving throughput.
                """);
        write(manuscriptDir.resolve("supplementary.tex"), """
                \\section*{Supplementary}
                Additional qualitative overlays are available in the results directory.
                """);
        write(manuscriptDir.resolve("references.bib"), """
                @article{kirillov2023segment,
                  title={Segment Anything},
                  author={Kirillov, Alexander and others},
                  journal={arXiv preprint arXiv:2304.02643},
                  year={2023}
                }
                """);
        write(manuscriptDir.resolve("manuscript.tex"), """
                \\documentclass{article}
                \\usepackage[margin=1in]{geometry}
                \\usepackage{graphicx}
                \\begin{document}
                \\title{SAMUeL: Segment Anything for Microscopy Using Labels}
                \\author{Automated report}
                \\maketitle
                \\input{introduction.tex}
                \\input{methods.tex}
                \\input{results.tex}
                \\input{discussion.tex}
                \\input{supplementary.tex}
                \\bibliographystyle{plain}
                \\bibliography{references}
                \\end{document}
                """);
        ProcessBuilder pb = new ProcessBuilder("pdflatex", "manuscript.tex");
        pb.directory(manuscriptDir.toFile());
        pb.redirectErrorStream(true);
        Process p = pb.start();
        p.waitFor();
    }

    private void write(Path path, String content) throws IOException {
        Files.writeString(path, content);
    }
}
